feat: add `lapack/base/dlartg`

aayush0325 opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds `lapack/base/dlartg`

## Related Issues

> Does this pull request have any related issues?

No. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

aayush0325 committed:
feat: add lapack/base/dlartg

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
feat: add main exports

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
docs: add ts files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
docs: add examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
bench: add benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
test: add tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
docs: update docs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
docs: add repl.txt

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlartg/README.md:
```diff
@@ -0,0 +1,292 @@
+<section class="intro">
+
+`dlartg` constructs a Givens rotation for a two-dimensional vector such that the following equation is satisfied:
+
```
```suggestion
This routine constructs a Givens rotation for a two-dimensional vector such that the following equation is satisfied:
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlartg/README.md:
```diff
@@ -0,0 +1,292 @@
+
+# dlartg
+
+> LAPACK routine to construct a Givens rotation of a two-dimensional vector that zeros out the second component.
```
@aayush0325 I'm a little confused. What is meant by a "two-dimensional" vector? Vectors are, by definition, one dimensional. Are you meaning a column matrix?

kgryte replied:
Or maybe you are intending a two-element vector?

aayush0325 replied:
i mean a column matrix here
<img width="811" height="207" alt="image" src="https://github.com/user-attachments/assets/4ed9f93c-6a7d-454c-93bb-3850a51dd60b" />
[ F ]
[ G ]

this is what im referring to as a 2 dimensional vector, i was a little confused here as well so i decided to go with this.

aayush0325 replied:
let me know if i should use something else here please i don't really know what else to use

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlartg/README.md:
```diff
@@ -0,0 +1,292 @@
+-   **G**: single element array representing the second component of the vector to be rotated as a [`Float64Array`][mdn-float64array].
+-   **C**: single element array overwritten by the cosine of the rotation as a [`Float64Array`][mdn-float64array].
+-   **S**: single element array overwritten by the sine of the rotation as a [`Float64Array`][mdn-float64array].
+-   **R**: single element array overwritten by the length of the rotated vector as a [`Float64Array`][mdn-float64array].
```
```suggestion
-   **F**: [`Float64Array`][mdn-float64array] containing a single element representing the first component of the vector to be rotated.
-   **G**: [`Float64Array`][mdn-float64array] containing a single element representing the second component of the vector to be rotated
-   **C**: [`Float64Array`][mdn-float64array] containing a single element which is overwritten by the cosine of the rotation.
-   **S**: [`Float64Array`][mdn-float64array] containing a single element which is overwritten by the sine of the rotation.
-   **R**: [`Float64Array`][mdn-float64array] containing a single element which is overwritten by the length of the rotated vector.
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlartg/README.md:
```diff
@@ -0,0 +1,292 @@
+
+The function has the following parameters:
+
+-   **F**: single element array representing the first component of the vector to be rotated as a [`Float64Array`][mdn-float64array].
```
Same comments as above.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlartg/README.md:
```diff
@@ -0,0 +1,292 @@
+The function has the following parameters:
+
+-   **F**: single element array representing the first component of the vector to be rotated as a [`Float64Array`][mdn-float64array].
+-   **offsetF**: starting index for `F`.
```
```suggestion
-   **offsetF**: index of the element in `F`.
```

It is somewhat awkward to say "starting index" when we are talking about only a single element. We can be more precise, IMO.

Applies here and below and likely elsewhere throughout this PR.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlartg/lib/base.js:
````diff
@@ -0,0 +1,116 @@
+* ```
+*
+* @private
+* @param {Float64Array} F - single element array representing the first component of the vector to be rotated
````
For the descriptions, all the same comments.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlartg/lib/base.js:
```diff
@@ -0,0 +1,116 @@
+var dlamch = require( '@stdlib/lapack/base/dlamch' );
+var sqrt = require( '@stdlib/math/base/special/sqrt' );
+var abs = require( '@stdlib/math/base/special/fast/abs' );
+var sign = require( '@stdlib/math/base/special/copysign' );
```
Prefer the alias `copysign`. `sign` means something else in other contexts.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlartg/lib/base.js:
```diff
@@ -0,0 +1,116 @@
+		R[ offsetR ] = sign( d, F[ offsetF ] );
+		S[ offsetS ] = G[ offsetG ] / R[ offsetR ];
+	} else {
+		u = min( safmax, max( safmin, f1, g1 ) );
```
Prefer two calls to fast `max`, rather than using `maxn`; e.g., `max( max( safmin, f1 ), g1 )`. Variadic functions are likely to end up in de-optimization hell. Here, we want to encourage compiler inlining.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlartg/lib/base.js:
```diff
@@ -0,0 +1,116 @@
+		u = min( safmax, max( safmin, f1, g1 ) );
+		fs = F[ offsetF ] / u;
+		gs = G[ offsetG ] / u;
+		d = sqrt( (fs*fs) + (gs*gs) );
```
This looks like `hypot`. Prefer that to avoid unnecessary underflow/overflow. I am surprised that LAPACK does not use `hypot` and/or the LAPACK equivalent routine.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlartg/lib/base.js:
```diff
@@ -0,0 +1,116 @@
+		S[ offsetS ] = sign( 1.0, G[ offsetG ] );
+		R[ offsetR ] = g1;
+	} else if ( f1 > rtmin && f1 < rtmax && g1 > rtmin && g1 < rtmax ) {
+		d = sqrt( (F[ offsetF ]*F[ offsetF ]) + (G[ offsetG ]*G[ offsetG ]) );
```
Same thing. This is `hypot`.

aayush0325 committed:
fix: use hypot and nested max in base implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
docs: update description

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---
